### PR TITLE
[Graph] Add API to convert graph to simple graph

### PR DESCRIPTION
**Pull request by jermainewang**
## Description

Add `dgl.to_simple_graph` to convert a graph to a simple graph. Also refactor the graph_index a little bit.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [Model], [Doc], [Feature]])
- [ ] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

## Changes
- [x] `dgl.to_simple_graph` API.
- [x] Refactor the `from_XXX` functions in `graph_index.py`. They are not class methods anymore. They take in the corresponding graph data and return a graph index object.
- [x] Double check all the places where `multigraph` flag is used, so that:
  * `multigraph` flag is default to be None (previously False). None means the flag will be determined by the data.
  * If `multigraph` flag is given by the user, the behavior is the same as before.
  * For mutable graph, the flag is default to be False (same behavior as before).

### CODE REVIEW

File: python/dgl/graph.py [hunk 1]
```diff
@@ -765,7 +765,8 @@ class DGLGraph(DGLBaseGraph):
     edge_frame : FrameRef, optional
         Edge feature storage.
     multigraph : bool, optional
-        Whether the graph would be a multigraph (default: False)
+        Whether the graph would be a multigraph. If none, the flag will be determined
+        by scanning the whole graph. (default: None)
```
Comment (zheng-da) [inline python/dgl/graph.py]: this sounds very expensive. then we should force users to provide `multigraph` for graph store.

Comment (zheng-da) [reply python/dgl/graph.py]: please remove the default value here. https://github.com/dmlc/dgl/blob/master/python/dgl/contrib/graph_store.py#L1020

Comment (jermainewang) [reply python/dgl/graph.py]: For immutable graph, the flag will be only computed once and only when `is_multigraph` is called. That'll be fine?

**Comment by jermainewang** [comment]
I found some issues with our CI. Will fix it in another PR.

**Comment by zheng-da** [comment]
I see there are many modifications in Jenkins. What are the modifications for?

File: python/dgl/graph.py (hunk 1 above)
Comment (zheng-da) [reply python/dgl/graph.py]: For giant graphs, I think it's better to ask users to provide this information? Is it fast to test it for a very large graph?

File: python/dgl/graph_index.py [hunk 2]
```diff
@@ -45,28 +46,29 @@ def __setstate__(self, state):
...
-            dst_ids.todgltensor(),
-            self._multigraph,
-            int(num_nodes),
-            self._readonly)
+            self._handle = _CAPI_DGLGraphCreate(
+                src.todgltensor(),
+                dst.todgltensor(),
+                multigraph,
+                int(num_nodes),
+                readonly)
```
Comment (zheng-da) [inline python/dgl/graph_index.py]: will it be better to use one API and the values of multigraph can be -1, 0, 1? too hacky?

File: src/graph/graph_op.cc [hunk 3]
```diff
@@ -273,4 +297,22 @@ IdArray GraphOp::ExpandIds(IdArray ids, IdArray offset) {
...
+      if (!hashmap.count(dst)) {
+        indices.push_back(dst);
+        hashmap.insert(dst);
+      }
+    }
+    indptr[src+1] = indices.size();
+  }
+  CSRPtr csr(new CSR(graph->NumVertices(), indices.size(),
+        indptr.begin(), indices.begin(), RangeIter(0), false));
+  return ImmutableGraph(csr);
```
Comment (zheng-da) [inline src/graph/graph_op.cc]: if the original graph index has only in_csr. the code first generates out_csr and return out_csr. Later on, the user's code needs in_csr. Maybe we don't need to spend time optimizing this operation. I just raise a potential issue. It's up to you to decide.

File: tests/graph_index/test_graph_index.py [hunk 4]
```diff
@@ -1,175 +0,0 @@
-from dgl import DGLError
```
Comment (zheng-da) [inline tests/graph_index/test_graph_index.py]: why removing this file?

Comment (jermainewang) [reply tests/graph_index/test_graph_index.py]: It looks a duplicate of `tests/compute/test_graph_index.py`.

**Comment by jermainewang** [comment]
> I see there are many modifications in Jenkins. What are the modifications for?

I'm testing the CI and found some bugs. I'll fix that in another PR and revert the changes here.

PLEASE DO NOT MERGE THIS PR

File: python/dgl/graph_index.py (hunk 2 above)
Comment (jermainewang) [reply python/dgl/graph_index.py]: That's fine as long as it is not exposed to user side. Will do that to eliminate some copy-pasting.

File: src/graph/graph_op.cc (hunk 3 above)
Comment (jermainewang) [reply src/graph/graph_op.cc]: Actually `SuccVec` will create `out_csr`, but you are right, it is always a decision to create which csr. I don't have good solution. It looks like a layout optimization problem that requires more information than simply one operator. For example, if we have a computation graph, then we know when and where to use which format so the format conversion cost + computation cost is minimized. Pretty far goal I will say.

File: python/dgl/graph.py (hunk 1 above)
Comment (jermainewang) [reply python/dgl/graph.py]: Yeah, we still allow user to directly provide the flag. I guess in the giant graph example, it's better to specify that explicitly. What do you think?

Comment (zheng-da) [reply python/dgl/graph.py]: yes, this is what I meaned.

File: python/dgl/graph_index.py (hunk 2 above)
Comment (jermainewang) [reply python/dgl/graph_index.py]: Done

**Pull request merged**